Core/Src/msb.c

```diff
@@ -7,151 +8,159 @@
 
 static osMutexAttr_t msb_i2c_mutex_attr;
 
-msb_t *init_msb(I2C_HandleTypeDef *hi2c, ADC_HandleTypeDef *adc1,
-		GPIO_TypeDef *debug_led1_gpio, uint16_t *debug_led1_pin,
-		GPIO_TypeDef *debug_led2_gpio, uint16_t *debug_led2_pin,
-		device_loc_t *device_loc)
-{
-	assert(hi2c);
-	assert(adc1);
-	assert(debug_led1_gpio);
-	assert(debug_led2_gpio);
-	assert(device_loc);
-
-	msb_t *msb = malloc(sizeof(msb_t));
-	assert(msb);
-
-	msb->hi2c = hi2c;
-	msb->adc1 = adc1;
-	msb->debug_led1_gpio = debug_led1_gpio;
-	msb->debug_led1_pin = debug_led1_pin;
-	msb->debug_led2_gpio = debug_led2_gpio;
-	msb->debug_led2_pin = debug_led2_pin;
-	msb->device_loc = device_loc;
+extern I2C_HandleTypeDef hi2c3;
+extern ADC_HandleTypeDef hadc1;
+extern device_loc_t device_loc;
+
+osMutexId_t i2c_mutex;
+
+#ifdef SENSOR_TEMP
+sht30_t temp_sensor;
+#endif
+
+#ifdef SENSOR_IMU
+lsm6dso_t imu;
+#endif
+
+#ifdef SENSOR_TOF
+VL6180xDev_t tof;
+#endif
 
+#if defined SENSOR_STRAIN || defined SENSOR_SHOCKPOT
+uint32_t adc1_buf[3];
+#endif
+
+int8_t init_msb()
+{
+#ifdef SENSOR_TEMP
 	/* Initialize the Onboard Temperature Sensor */
-	msb->temp_sensor = malloc(sizeof(sht30_t));
-	assert(msb->temp_sensor);
-	msb->temp_sensor->i2c_handle = hi2c;
-	assert(!sht30_init(msb->temp_sensor)); /* This is always connected */
+	temp_sensor = (sht30_t){
+		.i2c_handle = &hi2c3,
+	};
+	assert(!sht30_init(&temp_sensor)); /* This is always connected */
+#endif
 
+#ifdef SENSOR_IMU
 	/* Initialize the IMU */
-	msb->imu = malloc(sizeof(lsm6dso_t));
-	assert(msb->imu);
-	assert(!lsm6dso_init(msb->imu,
-			     msb->hi2c)); /* This is always connected */
+	imu = (lsm6dso_t){
 
-	/* Initialize the ToF sensor */
-	msb->tof = malloc(sizeof(VL6180xDev_t));
-	assert(msb->tof);
-	assert(!VL6180x_WaitDeviceBooted(msb->tof));
-	assert(!VL6180x_InitData(msb->tof));
-	assert(!VL6180x_Prepare(msb->tof));
+	};
+	assert(!lsm6dso_init(&imu, &hi2c3)); /* This is always connected */
+#endif
 
-	assert(!HAL_ADC_Start_DMA(msb->adc1, msb->adc1_buf,
-				  sizeof(msb->adc1_buf) / sizeof(uint32_t)));
+#ifdef SENSOR_TOF
+	/* Initialize the ToF sensor */
+	tof = malloc(sizeof(VL6180xDev_t));
+	assert(tof);
+	assert(!VL6180x_WaitDeviceBooted(tof));
+	assert(!VL6180x_InitData(tof));
+	assert(!VL6180x_Prepare(tof));
+#endif
+
+#if defined SENSOR_SHOCKPOT || defined SENSOR_STRAIN
```

Comment:
im not a million percent sure but doesnt it need to be defined(SENSOR_SHOCKPOT), i dont think the syntax of defined SENSOR_SHOCKPOT is valid

Reply:
They are itnerchangable. https://gcc.gnu.org/onlinedocs/cpp/Defined.html